src/core/ext/filters/client_channel/lb_policy/xds/xds_override_host.cc

```diff
@@ -646,79 +619,58 @@ XdsOverrideHostLb::SubchannelWrapper::SubchannelWrapper(
+  watcher_->watchers_.push_back(std::move(watcher));
 }
 
 void XdsOverrideHostLb::SubchannelWrapper::CancelConnectivityStateWatch(
     ConnectivityStateWatcherInterface* watcher) {
-  auto original_watcher = watchers_.find(watcher);
-  if (original_watcher != watchers_.end()) {
-    wrapped_subchannel()->CancelConnectivityStateWatch(
-        original_watcher->second);
-    watchers_.erase(original_watcher);
-  }
+  (void)std::remove_if(watcher_->watchers_.begin(), watcher_->watchers_.end(),
+                       [watcher](const auto& watcher_ptr) {
+                         return watcher_ptr.get() == watcher;
+                       });
 }
 
-//
-// factory
-//
-const JsonLoaderInterface* XdsOverrideHostLbConfig::JsonLoader(
-    const JsonArgs&) {
-  static const auto kJsonLoader =
-      JsonObjectLoader<XdsOverrideHostLbConfig>()
-          // Child policy config is parsed in JsonPostLoad
-          .Finish();
-  return kJsonLoader;
+grpc_pollset_set* XdsOverrideHostLb::SubchannelWrapper::
+    ConnectivityStateWatcher::interested_parties() {
+  if (interested_parties_ != nullptr) {
+    grpc_pollset_set_destroy(interested_parties_);
+    interested_parties_ = nullptr;
+  }
+  interested_parties_ = grpc_pollset_set_create();
+  for (const auto& watcher : watchers_) {
+    grpc_pollset_set_add_pollset_set(interested_parties_,
+                                     watcher->interested_parties());
+  }
+  return interested_parties_;
 }
 
-void XdsOverrideHostLbConfig::JsonPostLoad(const Json& json, const JsonArgs&,
-                                           ValidationErrors* errors) {
-  {
-    ValidationErrors::ScopedField field(errors, ".childPolicy");
-    auto it = json.object_value().find("childPolicy");
-    if (it == json.object_value().end()) {
-      errors->AddError("field not present");
-    } else {
-      auto child_policy_config = CoreConfiguration::Get()
-                                     .lb_policy_registry()
-                                     .ParseLoadBalancingConfig(it->second);
-      if (!child_policy_config.ok()) {
-        errors->AddError(child_policy_config.status().message());
-      } else {
-        child_config_ = std::move(*child_policy_config);
-      }
-    }
-  }
-  {
-    ValidationErrors::ScopedField field(errors, ".overrideHostStatus");
-    auto it = json.object_value().find("overrideHostStatus");
-    if (it != json.object_value().end()) {
-      override_host_status_mask_ = 0;
-      auto array = it->second.array_value();
-      for (const auto& value : array) {
-        auto status = XdsHealthStatus::FromString(value.string_value());
-        if (status.has_value()) {
-          override_host_status_mask_ |= HealthStatusBitMask(*status);
-        }
-      }
-    }
+void XdsOverrideHostLb::SubchannelWrapper::ConnectivityStateWatcher::
+    OnConnectivityStateChange(grpc_connectivity_state state,
+                              absl::Status status) {
+  connectivity_state_ = state;
```

Comment:
Please use `connectivity_state_.store(state)`, so that it's clear to the reader that this is an atomic.